src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Encoder.cs

```diff
@@ -226,6 +235,82 @@ public static unsafe OperationStatus EncodeToUtf8InPlace(Span<byte> buffer, int
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Avx512BW))]
+        [CompExactlyDependsOn(typeof(Avx512Vbmi))]
+        private static unsafe void Avx512Encode(ref byte* srcBytes, ref byte* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, byte* destStart)
+        {
+            // Reference for VBMI implementation : https://github.com/WojciechMula/base64simd/tree/master/encode
+            // If we have AVX512 support, pick off 48 bytes at a time for as long as we can.
+            // But because we read 64 bytes at a time, ensure we have enough room to do a
+            // full 64-byte read without segfaulting.
+
+            byte* src = srcBytes;
+            byte* dest = destBytes;
+
+            // The JIT won't hoist these "constants", so help it
+            // JIT will remove the Non VBMI path constants when VBMI is available and vice-versa.
+            Vector512<sbyte> shuffleVecVbmi = Vector512.Create(
+                0x01020001, 0x04050304, 0x07080607, 0x0a0b090a,
+                0x0d0e0c0d, 0x10110f10, 0x13141213, 0x16171516,
+                0x191a1819, 0x1c1d1b1c, 0x1f201e1f, 0x22232122,
+                0x25262425, 0x28292728, 0x2b2c2a2b, 0x2e2f2d2e).AsSByte();
+            Vector512<sbyte> vbmiLookup = Vector512.Create("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"u8).AsSByte();
+
+            Vector512<ushort> maskAC = Vector512.Create((uint)0x0fc0fc00).AsUInt16();
+            Vector512<uint> maskBB = Vector512.Create((uint)0x3f003f00);
+            Vector512<ushort> shiftAC = Vector512.Create((uint)0x0006000a).AsUInt16();
+            Vector512<ushort> shiftBB = Vector512.Create((uint)0x00080004).AsUInt16();
+
+            AssertRead<Vector256<sbyte>>(src, srcStart, sourceLength);
+
+            // This algorithm requires AVX512VBMI support.
+            // Vbmi was first introduced in CannonLake and is avaialable from IceLake on.
+            // This makes it okay to use Vbmi instructions since Vector512.IsHardwareAccelerated returns True only from IceLake on.
+
+            // str = [...|PONM|LKJI|HGFE|DCBA]
+            Vector512<sbyte> str = Vector512.Load(src).AsSByte();
+
+            while (true)
```

Comment:
Could this be written as a do-while loop instead to avoid duplicating the read logic?

Reply:
> Could this be written as a do-while loop instead to avoid duplicating the read logic?

Decided to stick to similar structure to avx2Encode - only the specific instructions used vary. But open to considering this

Reply:
Looks like in the case of `Avx2Encode`, the first load is actually different, so it makes sense to move it out of the loop there.

But it doesn't matter much either way, just a minor style preference since it removes a few redundant lines.